### DISCUSSION TITLE
Avalonia and scatterplot (using scottplot)

### DISCUSSION THREAD

**Discussion by mijenner** [AvaloniaUI/Avalonia#8059] [Q&A] [answered]
Newbie question: how to produce simple scatterplot using ScottPlot and Avalonia? 

I think I'm close to a solution and at the same time extremely far away from the solution - in terms of understanding the essentials. 

My Model is a state (Tilstand.cs) holding three doubles, time (elapsed since simulation started), position and velocity. App is called Fritfaldssim and Model is located in Models folder: 

```
namespace Fritfaldssim.Models
{
   public class Tilstand
   {
      public double timestamp;
      public double position;
      public double speed; 
   }
}
```

Then I have service Simulate.cs which generates the data, later to be plotted. It includes below property:  

```
public ObservableCollection<Tilstand> SimResult { get;  }
```

I try to reference above data in the MainWindowViewModels.cs through it's constructor (don't want to add extra view models, the template generated should be sufficient for this small app). 

```
public ObservableCollection<Tilstand>? SimResultsList { set;  get; }

public MainWindowViewModel(ObservableCollection<Tilstand> sim)
{
   SimResultsList = sim;
}
```

And in the OnFrameworkInitializationCompleted() method in App.axaml.cs I instantiate the service Simulate, and feed a reference to MainWindowViewModel through its constructor. 

```
sim = new Simulate(); 
desktop.MainWindow = new MainWindow
{
   DataContext = new MainWindowViewModel(sim.SimResult),
};
```

However, in MainWindow.axaml.cs I cannot get hold of the data and feed it to scottplot's Plot.AddScatter() method. 

Hints, solutions, suggestions are very welcomed! A quick and dirty (and easy to implement) solution is probably better for me than a perfectly structured solution. 

The solution / project was developed in Visual Studio 2022 and is available here: [https://github.com/mijenner/Fritfaldssim](https://github.com/mijenner/Fritfaldssim)

**Comment by timunie** [comment]
I don't know much about scot plot, but the properties in you model may be missing the get accessor. At least for all other bindings in Avalonia you need to have a get; accessor. 

```cs
namespace Fritfaldssim.Models
{
   public class Tilstand
   {
      public double timestamp {get; private set;}
      public double position {get; private set;}
      public double speed {get; private set;}
   }
}
```

**Reply by mijenner** [reply]
I actually wasn't using the data from the Models directly. Instead I create them in the Simulate "service", and reference these data from a property, with get, in MainWindowViewModel. 

I've fixed some other bugs, and changed format of stored data to make it easier to provide two arrays, e.g. time and speed for scottplot, but still no success. 

Other hints are very welcomed!

**Comment by timunie** [comment, marked as answer]
Ok I have had a deeper look today. Seems like your DataContext is null in the Constructor. Therefore I listened to DataContextChanged and once the datacontext is set, update the plot. This seems to work fine. for me: https://github.com/mijenner/Fritfaldssim/pull/1

preview: 
![image](https://user-images.githubusercontent.com/47110241/166197903-4c22d102-9d6b-4242-9e29-a5924b20e507.png)

If you want to update the plot dynaimcally, I think you need to find a way to subscribe to property changes and update the plot then. 

Happy coding
Tim

**Reply by mijenner** [reply]
Cool, thanks! 

Later, I would like to know why the DataContext was null initially, and first set later. But your solution was spot on, thanks! 

Kind regards, 
Michael
